examples/DivisorCurve.jl

```diff
@@ -0,0 +1,327 @@
+
+function help_int_multiplication(D::CurveDivisor, k::Int)
+    F = copy(D.divisor)
+    for P in keys(F)
+        F[P] = k*D.divisor[P]
+    end
+    return F
+end
+
+function Base.:*(k::Int, D::AffineCurveDivisor)
+    F = help_int_multiplication(D, k)
+    return AffineCurveDivisor(D.C, F)
+end
+
+function Base.:*(k::Int, D::ProjCurveDivisor)
+    F = help_int_multiplication(D, k)
+    return ProjCurveDivisor(D.C, F)
+end
+
+################################################################################
+
+function ==(D::CurveDivisor, E::CurveDivisor)
+    return D.C == E.C && D.divisor == E.divisor
+end
+
+################################################################################
+################################################################################
+
+@doc Markdown.doc"""
+    iseffective(D::CurveDivisor)
+
+Return `true` if `D` is an effective divisor, `false` otherwise.
+"""
+function iseffective(D::CurveDivisor)
+    return all(v >= 0 for v in values(D.divisor))
+end
+
+################################################################################
+################################################################################
+
+@doc Markdown.doc"""
+    multiplicity(C::AffinePlaneCurve, phi::AbstractAlgebra.Generic.Frac, P::Point)
+
+Return the multiplicity of `phi` on the curve `C` at the point `P`.
+"""
+function multiplicity(C::AffinePlaneCurve, phi::AbstractAlgebra.Generic.Frac, P::Point)
+    f = divrem(phi.num, C.eq)
+    g = divrem(phi.den, C.eq)
+    !iszero(g[2]) || error("This is not a rational function on `C`.")
+    if iszero(f[2])
+        return Inf
+    elseif isconstant(f[2])
+        a = 0
+    else
+        a = intersection_multiplicity(C, AffinePlaneCurve(f[2]), P)
+    end
+    if isconstant(g[2])
+        b = 0
+    else
+        b = intersection_multiplicity(C, AffinePlaneCurve(g[2]), P)
+    end
+    return a-b
+end
+
+###############################################################################
+
+@doc Markdown.doc"""
+    multiplicity(C::AffinePlaneCurve{S}, F::Oscar.MPolyElem{S}, P::Point{S}) where S <: FieldElem
+
+Return the multiplicity of `F` on the curve `C` at the point `P`.
+"""
+function multiplicity(C::AffinePlaneCurve{S}, F::Oscar.MPolyElem{S}, P::Point{S}) where S <: FieldElem
+    f = divrem(F, C.eq)
+    if iszero(f[2])
+        return Inf
+    elseif isconstant(f[2])
+        return 0
+    else
+        return intersection_multiplicity(C, AffinePlaneCurve(f[2]), P)
+    end
+end
+
+################################################################################
+
+@doc Markdown.doc"""
+    multiplicity(C::ProjPlaneCurve{S}, F::Oscar.MPolyElem{S}, P::Oscar.Geometry.ProjSpcElem{S}) where S <: FieldElem
+
+Return the multiplicity of `F` on the curve `C` at the point `P`.
+"""
+function multiplicity(C::ProjPlaneCurve{S}, F::Oscar.MPolyElem{S}, P::Oscar.Geometry.ProjSpcElem{S}) where S <: FieldElem
+    f = divrem(F.f, defining_equation(C))
```

Comment:
For the affine case we apparently have to write `f = divrem(F, C.eq)` here. I find this discrepancy unsettling; what is the cause for this? I am concerned it may be a source of errors down the road. Unifying this (and a few other hopefully minor things) might also allow us to reduce code duplication between affine and projective cases in the future.

Reply:
I made a mistake in the input, F was supposed to be a decorated polynomial, and I need to use F.f to get access to the underlying non-decorated polynomial below F.